components/servo/webview.rs

```diff
@@ -7,21 +7,149 @@ use std::ffi::c_void;
+    // unsure what uses this and unsure of its semantics
+    // fn web_resource_requested(_: WebView, _: WebResourceRequest, _: IpcSender<WebResourceResponseMsg>);
+    /// A pipeline in the webview panicked. First string is the reason, second one is the backtrace.
+    fn notify_crashed(_: WebView, _: String, _: Option<String>);
+    /// Open dialog to select bluetooth device.
+    fn show_bluetooth_device_dialog(_: WebView, _: Vec<String>, _: IpcSender<Option<String>>);
+    /// Open file dialog to select files. Set boolean flag to true allows to select multiple files.
+    fn show_file_selection_dialog(
+        _: WebView,
+        _: Vec<FilterPattern>,
+        _: bool,
+        _: IpcSender<Option<Vec<String>>>,
+    );
+    /// Open interface to request permission specified by prompt.
+    fn request_permission(_: WebView, _: PermissionPrompt, _: IpcSender<PermissionRequest>);
+    /// Request to present an IME to the user when an editable element is focused.
+    /// If the input is text, the second parameter defines the pre-existing string
+    /// text content and the zero-based index into the string locating the insertion point.
+    /// bool is true for multi-line and false otherwise.
+    fn show_ime(
+        _: WebView,
+        _: InputMethodType,
+        _: Option<(String, i32)>,
+        _: bool,
+        _: DeviceIntRect,
+    );
+    /// Request to hide the IME when the editable element is blurred.
+    fn hide_ime(_: WebView);
+    /// Report a complete sampled profile
+    // fn report_profile(_: Vec<u8>); // do we still need this?
+    /// Notifies the embedder about media session events
+    /// (i.e. when there is metadata for the active media session, playback state changes...).
+    fn notify_media_session_event(_: WebView, _: MediaSessionEvent);
+    /// Notify the embedder that it needs to present a new frame.
+    fn notify_new_frame_ready(_: WebView, _: Vec<WebViewId>);
+    /// The given event was delivered to a pipeline in the given webview.
+    fn notify_event_delivered(_: WebView, _: CompositorEventVariant);
+    /// Request to play a haptic effect on a connected gamepad.
+    fn play_gamepad_haptic_effect(
+        _: WebView,
+        _: usize,
+        _: GamepadHapticEffectType,
+        _: IpcSender<bool>,
+    );
+    /// Request to stop a haptic effect on a connected gamepad.
+    fn stop_gamepad_haptic_effect(_: WebView, _: usize, _: IpcSender<bool>);
+}
+
+pub trait ServoDelegate {
+    /// Servo has shut down
+    fn notify_shutdown();
+    /// Report the status of Devtools Server with a token that can be used to bypass the permission prompt.
+    fn notify_devtools_server_started(_: Result<u16, ()>, _: String);
+}
+
+pub trait WeDidntGetTimeForThisDelegateButItWasFormerlyWindowMethods {
+    fn get_coordinates(&self) -> compositing::windowing::EmbedderCoordinates;
+    fn notify_animation_state_changed(&self, state: compositing::windowing::AnimationState);
+}
```

Comment:
Here's the discussion that animation state change can move to servo delegate. I would expect both stateful and immediate mode would like to keep polling when the state change to animating.

`EmbedderCoordinates`, however, is more tricky. I hope we can remove it and find out what exact values they want from each call.